### PR TITLE
Feature/pdct 1559 download all data should only show data from allowed corpora

### PR DESCRIPTION
**Pull request by katybaulch**
# Description

- Restrict download this search data to only data in allowed corpora
- Restrict download all data to only data in allowed corpora
- Always exclude MCF from download
- Add test coverage for this search and whole data download
- Refactor app token factory into class
- Update existing vespa search tests
- Use LRU cache for whole database download query

## Proposed version

Please select the option below that is most relevant from the list below. This
will be used to generate the next tag version name during auto-tagging.

- [ ] Skip auto-tagging
- [x] Patch
- [ ] Minor version
- [ ] Major version

Visit the [Semver website](https://semver.org/#summary) to understand the
difference between `MAJOR`, `MINOR`, and `PATCH` versions.

Notes:

- If none of these options are selected, auto-tagging will fail
- Where multiple options are selected, the most senior option ticked will be
  used -- e.g. Major > Minor > Patch
- If you are selecting the version in the list above using the textbox, make
  sure your selected option is marked `[x]` with no spaces in between the
  brackets and the `x`

## Type of change

Please select the option(s) below that are most relevant:

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] GitHub workflow update
- [ ] Documentation update
- [ ] Refactor legacy code
- [ ] Dependency update

## How Has This Been Tested?

Please describe the tests that you added to verify your changes.

## Reviewer Checklist

- [ ] **DB_CLIENT DEPENDENCY IS ON THE LATEST VERSION**
- [ ] The PR represents a single feature (small driveby fixes are also ok)
- [ ] The PR includes tests that are sufficient for the level of risk
- [ ] The code is sufficiently commented, particularly in hard-to-understand areas
- [ ] Any required documentation updates have been made
- [ ] Any TODOs added are captured in future tickets
- [ ] No FIXMEs remain

### CODE REVIEW

File: app/api/api_v1/routers/search.py [hunk 1]
```diff
@@ -203,22 +174,45 @@ def search_documents(
...
+                "app_token": str(app_token),
             }
         },
     )
+
+    # Decode the app token and validate it.
+    #
+    # First corpora validation is app token against DB. At least one of the app token
+    # corpora IDs must be present in the DB to continue the search request.
+    token = AppTokenFactory()
```
Comment (diversemix) [inline app/api/api_v1/routers/search.py:197]: COMMENT: You may well end up writing some middleware to do this

**Review by diversemix — APPROVED** [review]
lgtm ... general comment for post-COP (you probably know this) the routers are getting far too much code in at the moment.

File: app/api/api_v1/routers/search.py (hunk 1 above)
Comment (odrakes-cpr) [reply app/api/api_v1/routers/search.py:197]: @diversemix what do you mean here exactly?

Comment (katybaulch) [reply app/api/api_v1/routers/search.py:197]: Yeah - totally agree!

**Pull request merged**